### ISSUE TITLE
✂Cut  inside/outside elements

### ISSUE THREAD

**Issue by dnknn** [gildas-lormeau/SingleFile#516]
<details>
<summary>Details</summary>

![2010_15203403](https://user-images.githubusercontent.com/10995302/96123710-cc41c100-0f25-11eb-8126-f45d23be745b.png)
Cut  outside elements, It’s really great, this feature is very, very good, completely beyond my personal expectations.  Now I have a few ideas to strengthen it, WDYT？

</details>

----

<details>
<summary><s>Done</summary>

- #   Add cut outside elements of default mode.
/options.html -> Annotation editor -> default mode
![2010_15204156](https://user-images.githubusercontent.com/10995302/96127123-e334e300-0f26-11eb-9d8c-d31be6268290.png)
Cannot select outside-cut mode at present. </s>

</details>

----

<details>
<summary><s>Done</summary>

- # Multi-select operation support.
now, `Left-clicking` or `Space` cut/remove selected-outside-elements , no any change.
then, additional `Ctrl+Left-clicking`  to keep Multi-selected .
`Ctrl+clicking` is just for keep Multi-selected and will not perform any operation.
The selected element, keep the outline style (the other CSS style can also be added separately).
At last, use `Left-clicking` or `Space` can cut/remove all selected-outside-elements.

![2010_15211431](https://user-images.githubusercontent.com/10995302/96131173-6e17dc80-0f2b-11eb-90ef-a377217c1006.png)

remove/cut  Multi-selected-outside-elements, this can greatly improve the practicality.
The most typical example, 
![2010_15213505](https://user-images.githubusercontent.com/10995302/96136240-52fa9c00-0f2e-11eb-98d8-fca6415cdfcd.png)
Suppose there is an element of an article, including the title(A) and content(B) and the sidebar(C), three sub-elements.
And this function allows you to remove any elements other than A and B.
</s>

</details>

---

<details>
<summary><s>Done</summary>

- # Reverse inside➰outside cut/remove.

Use additional key to **temporary** reverse inside/outside,  Note that it is **one time temporary**, not switching cut-mode. 
I recommend `Shift` for this additional  key. which is: `Shift+Space` or `Shift+clicking`.

This function can be used in inside-cut mode and outside-cut mode,  but in outside-cut mode, it is most useful, it can allows you to temporarily reverse.
![2010_15220133](https://user-images.githubusercontent.com/10995302/96139997-02853d80-0f32-11eb-9f12-8863f709f7cc.png) </s>

</details>

**Comment by dnknn** [comment]
2 and 3 , I think you should be able to understand the specific meaning.

**Comment by gildas-lormeau** [comment]
Thank you for your suggestions. I think 1 is 2 are ok, I'm not sure about 3. What should happen if the user presses the shift key when he is selecting several elements?

**Comment by dnknn** [comment]
> I'm not sure about 3. What should happen if the user presses the shift key (eventually several times) when he is selecting several elements?

It's very simple, the additional `Shift` key does not work for "when he is selecting several elements".
In other words, it is only used to reverse a single element-select, for multiple reversals, it is meaningless in itself.

Now you understand the specific meaning? 

In a word:
Use `Shift+Space` or `Shift+clicking` to reverse between `inside-cut` and `outside-cut`, it is only suitable for one-time temporary reverse of **a single element**.

Its purpose:  very useful.

![2010_16180800](https://user-images.githubusercontent.com/10995302/96245967-8b57b400-0fda-11eb-9fd3-a5de4aaa8182.png)
It allows you to reverse directly without switching ❶or❷ mode.

**Comment by dnknn** [comment]
@gildas-lormeau 

For 2, Multi-select operation support,  `Ctrl+clicking`, It is useful in outside-cut-mode.  

> I'm not sure about 3. What should happen if the user presses the shift key (eventually several times) when he is selecting several elements?

It's very simple, the additional `Shift` key does not work for "when he is selecting several elements".
In other words, it is only used to reverse a single element-select, 
### because for multiple-element-select reversals, it is meaningless in itself.

and because It means **Reverse** / **Transform** /**Conversion** , Therefore, it is completely unnecessary to work on multiple elements, otherwise it will cause confusion and incomprehension.

Now you understand the specific meaning? 

One sentence:
Use `Shift+Space` or `Shift+clicking` to reverse between `inside-cut` and `outside-cut`, it is only suitable for one-time temporary reverse of **a single element**.

Its purpose:  very useful.

![2010_16180800](https://user-images.githubusercontent.com/10995302/96245967-8b57b400-0fda-11eb-9fd3-a5de4aaa8182.png)
It allows you to reverse directly without switching ❶or❷ mode.

---
Example: If the selector of the currently selected element is `#abc`
> I just give an example, the code syntax is not necessarily accurate.

### for in inside-cut-mode: 
```
#abc {display:none !important}
```

if  `Shift+Space` or `Shift+clicking` It will become:
```
:not(#abc) {display:none !important}
````

### for in outside-cut-mode: 
```
:not(#abc) {display:none !important}
```
if  `Shift+Space` or `Shift+clicking` It will become:
```
#abc {display:none !important}
```

As can be seen, It means **Reverse** / **Transform** /**Conversion** 
but note that it only applies to the reverse operation of a single element.

**Comment by gildas-lormeau** [comment]
Thank you for the additional details. The features you suggested are implemented. Feel free to test them and report the issues you might find.

**Comment by dnknn** [comment]
I tested it and it fully met expectations. very great, thank you!  and I report two issues:

### Problems with `Shift+clicking`
`Shift+clicking` will be treated as `text-selection` at the same time.
The gif also demo: normal default text-selection for `Shift+clicking`.
![bug007](https://user-images.githubusercontent.com/10995302/96333864-071b3480-109f-11eb-99c7-3a214337b5eb.gif)

　　

---
### Add to support `Esc` to exit multi-selection.

![Esc](https://user-images.githubusercontent.com/10995302/96333969-d12a8000-109f-11eb-8d72-9cb1aa3c7f77.gif)

**Comment by dnknn** [comment]
@gildas-lormeau 

I tested it and it fully met expectations. very great, thank you!  and I report two issues:

- ### Problems with `Shift+clicking`
`Shift+clicking` will be treated as `text-selection` at the same time.
The gif also demo: normal default text-selection for `Shift+clicking`.
![bug007](https://user-images.githubusercontent.com/10995302/96333864-071b3480-109f-11eb-99c7-3a214337b5eb.gif)

---
- ### Problems with `Shift+clicking` / `Shift+Space` under multi-selection.
Under multi-selection, `Shift+clicking`/ `Shift+Space` will be treated as `clicking`/ `Space`.
You still don't fully understand the meaning,  I think `Shift+clicking`/ `Shift+Space` is only for a single instant :hover element, regardless of whether it is multi-selection.
For example: You multi-selected the 3 elements A B C.
Then `Shift+clicking`/ `Shift+Space` A can individually reverse A, and `Shift+clicking`/ `Shift+Space` D can individually reverse D. 
In other words, it doesn’t care about multi-selection, It only reverses a single element, not multiple elements.

---
Add to support `Esc` to exit multi-selection.  It can be omitted. this is not important, I'm just talking about it.  Because Ctrl+clicking can easily exit a multiple selection again.

<details>
<summary>Details</summary>

![Esc](https://user-images.githubusercontent.com/10995302/96333969-d12a8000-109f-11eb-8d72-9cb1aa3c7f77.gif)

</details>

**Comment by gildas-lormeau** [comment]
Thank you, I fixed the issues related to the Shift key. I decided to block the text selection when the user is cutting elements. I also added the support of the Escape key to reset the selection of multiple elements.

**Comment by dnknn** [comment]
Thanks♪(･ω･)ﾉ @gildas-lormeau 
> I decided to block the text selection when the user is cutting elements.

+1👍 Very wise!!!

### Report a new 🐞:   Cannot use Ctrl+Z/Ctrl+Shift+Z.
> I deleted the old version and tested the latest repo that I downloaded, and then I found a new bug (the repo yesterday seem did not have this bug)

### Conditions to reproduce this bug:  see ↓
![2010_18054908](https://user-images.githubusercontent.com/10995302/96354128-aa069980-1105-11eb-8fcb-114af9e99f91.png)

❶ : 　　no problem!
❷or❸:  Enter annotate the page for the **first** time, Cannot Ctrl+Z/Ctrl+Shift+Z.

>  Note that it is the first time to enter.

Use buttons are works(👇).  Press the button several times, and Surprised it's: can Ctrl+Z/Ctrl+Shift+Z to work again.
![2010_18055156](https://user-images.githubusercontent.com/10995302/96354165-0b2e6d00-1106-11eb-8eaf-082cf1c4875e.png)

**Comment by gildas-lormeau** [comment]
Thank you! It should be fixed now.

**Comment by dnknn** [comment]
> Thank you! It should be fixed now.

Yes. nice! fixed! 

Last question, have you thought about it ?
If the inline style or CSS contains: ` pointer-events:none; `
The element of `pointer-events:none` means: You cannot `:hover` it at this time, that is, you cannot select element. right?
So, does SingleFile's removing mode need to be additionally reversed to achieve element selection?
You can add `pointer-events:none` CSS to test by yourself.

**Comment by gildas-lormeau** [comment]
Good point! it's also fixed now.

**Comment by dnknn** [comment]
> Good point! it's also fixed now.

awesome!      so, I close this issue and give you a good suggestion.

https://github.com/gildas-lormeau/SingleFile/commit/23d47110d866b0e3d9de40cb97e1a30ff5fd0517 added Ctrl+S, very very good! 

Then I suggest you add the hotkey directly to the DOM `title`

`title="Ctrl+S: Save the page"`
`title="Ctrl+Shift+Z: Remove last restored element"`
...
![2010_18205657](https://user-images.githubusercontent.com/10995302/96368097-79197980-1184-11eb-9445-fa5e1864542c.png)

This way, it is very good for users.

**Comment by dnknn** [comment]
> title="Ctrl+Shift+Z: Remove last restored element"
title="Ctrl+Z: ...."

Supplementary correction: It seems not quite right,  Because they are not suitable for global mode.

But title="Ctrl+S: Save the page" is ok

**Comment by gildas-lormeau** [comment]
There are delimiters between groups of buttons that are supposed to help the user understand that they are not global.
